Single-File: Pass BUNDLE_PROBE property to the runtime

swaroop-sridhar opened the pull request:
As described in the [design doc](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#startup), pass the bundle_probe function pointer encoded as a string to the runtime.

swaroop-sridhar committed:
Single-File: Pass BUNDLE_PROBE property to the runtime

As described in the [design doc](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#startup), pass the bundle_probe function pointer encoded as a string to the runtime.

vitek-karas commented on src/installer/corehost/cli/bundle/runner.cpp:
```diff
@@ -81,3 +81,47 @@ bool runner_t::locate(const pal::string_t& relative_path, pal::string_t& full_pa
+    return false;
+}
+
+pal::string_t runner_t::get_bundle_probe()
```
Just my personal preference: I would move this to hostpolicy coreclr setup code. I don't think the runner is the right component to know about who we encode the probe value to pass it to the runtime - that's a concern which is solved by hostpolicy coreclr interaction code.

vitek-karas commented on src/installer/corehost/cli/bundle/runner.h:
```diff
@@ -37,8 +37,22 @@ namespace bundle
+        // * The runtime uses Unicode strings on all platforms
+        // Using a unicode encoded path presents a uniform interface to the runtime
+        // and minimizes the number if Unicode <-> UTF8 conversions necessary.
+        static bool __stdcall bundle_probe(const wchar_t* path, int64_t* offset, int64_t* size);
```
I would make this a top-level/global function and not a member function - to make it clear it's effectively an exported API. Or maybe even better - do the same which hostpolicy already does for functions like `get_delegate` and similar. Those are also effectively exported APIs which are returned as function pointers. Currently they are written as hidden inside an anonymous namespace and don't need header/cpp duplicate declarations.

Also currently the header and cpp have slightly different declarations of the function - which feels wrong.

vitek-karas commented on src/installer/corehost/cli/hostmisc/pal.unix.cpp:
```diff
@@ -13,9 +13,12 @@
 #include <fcntl.h>
 #include <fnmatch.h>
 #include <ctime>
+#include <clocale>
```
Can you please make sure that this change doesn't regress the size of apphosts? In theory it should not as the new functionality is not used by apphosts, but still.

swaroop-sridhar replied:
On Linux X64 Release build, apphost binary included the `unicode_palstring` function -- so it wasn't removed by the compiler. But the apphost size only increased by 16 bytes (the function's size is definitely bigger than 16 bytes).

vitek-karas reviewed (approved):
Just minor stuff, looks good.

swaroop-sridhar committed:
Address feedback by @vitek-katas

swaroop-sridhar committed:
bundle_probe() use char16_t* string type

Type the path-string as a char16_t* string instead of wchar_t*
wchar_t* size varies by architecture.

bool STDMETHODCALLTYPE bundle_probe(const char16_t* path, int64_t* offset, int64_t* size);

swaroop-sridhar commented:
@vitek-karas  I updated the `bundle_probe` interface from 
```
bool STDMETHODCALLTYPE bundle_probe(const wchar_t* path, int64_t* offset, int64_t* size)
```
to 
```
bool STDMETHODCALLTYPE bundle_probe(const char16_t* path, int64_t* offset, int64_t* size)
```
because:

* `wchar_t` is 16-bit encoding on Windows while it is 32-bit encoding on most Unix systems
* The runtime uses 16-bit encoded unicode characters.

Therefore, the interface uses the fixed char16_t type. This fixed the test failures on Linux.
